### PR TITLE
Wacodis developments

### PR DESCRIPTION
**Pull request by matthesrieke**
this PR brings in some small fixes for issues we came across during the wacodis deployments.

### CODE REVIEW

**Comment by matthesrieke** [comment]
thankfully, checkstyle exists................... ........

**Comment by autermann** [comment]
There has to be a better solution. Even a `ReadWriteLock` instead of the synchronized block would be way faster. Right now each and every call to  `getStatus(JobId)` is exclusive and no job can complete (or it's status can be requested) while the result of about any other job is persisted. And that may take some time.

**Comment by matthesrieke** [comment]
@autermann good point, see my new commit. There are very likely other workflow patterns, which are faster (e.g. eventual consistency via an async/bus-based persistence), but this would be a bigger re-engineering. This fix addresses a real world issue (coming from the auto-polling feature of the wps-client-lib) and resolve the problem in a pragmatic way. We could consider a refactor/redesign later on.

**Comment by bpross-52n** [comment]
Bitte nochmal Bescheid geben, wenn der PR reviewed werden soll.

**Comment by matthesrieke** [comment]
@bpross-52n gerne reviewen, momentan sind keine weiteren Anpassungen geplant. thx!

File: engine/src/main/java/org/n52/javaps/engine/impl/EngineImpl.java [hunk 1]
```diff
@@ -138,20 +140,28 @@ public StatusInfo dismiss(JobId identifier) throws JobNotFoundException {
...
+        StatusInfo result;
         if (job != null) {
-            return job.getStatus();
+            result = job.getStatus();
         } else {
-            return this.resultPersistence.getStatus(identifier);
+            result = this.resultPersistence.getStatus(identifier);
         }
+
+        this.jobStatusLock.readLock().unlock();
```
Comment (autermann) [inline engine/src/main/java/org/n52/javaps/engine/impl/EngineImpl.java]:
```java
this.jobStatusLock.readLock().lock();
try {
  ...
} finally {
  this.jobStatusLock.readLock().unlock();
}
```

File: engine/src/main/java/org/n52/javaps/engine/impl/EngineImpl.java [hunk 2]
```diff
@@ -182,20 +192,25 @@ public JobId execute(OwsCode identifier,
 
     private Result onJobCompletion(Job job) throws EngineException {
         try {
+            this.jobStatusLock.writeLock().lock();
```
Comment (autermann) [inline engine/src/main/java/org/n52/javaps/engine/impl/EngineImpl.java]: you should lock it before the `try`

File: engine/src/main/java/org/n52/javaps/engine/impl/EngineImpl.java [hunk 3]
```diff
@@ -182,20 +192,25 @@ public JobId execute(OwsCode identifier,
...
             this.cancelers.remove(job.getJobId());
             this.resultPersistence.save(job);
             this.jobs.remove(job.getJobId());
             return this.resultPersistence.getResult(job.getJobId());
         } finally {
             job.destroy();
+            this.jobStatusLock.writeLock().unlock();
```
Comment (autermann) [inline engine/src/main/java/org/n52/javaps/engine/impl/EngineImpl.java]: If `job.destroy()` fails, the lock stays locked...

File: engine/src/main/java/org/n52/javaps/engine/impl/EngineImpl.java [hunk 4]
```diff
@@ -182,20 +192,25 @@ public JobId execute(OwsCode identifier,
...
         }
     }
 
     @Override
     public Future<Result> getResult(JobId identifier) throws JobNotFoundException {
         LOG.info("Getting result {}", identifier);
+
+        this.jobStatusLock.readLock().lock();
         Job job = this.jobs.get(identifier);
```
Comment (autermann) [inline engine/src/main/java/org/n52/javaps/engine/impl/EngineImpl.java]: Same as above, if `this.jobs.get(identifier);`, fails, the lock is never unlocked.

File: engine/src/main/java/org/n52/javaps/engine/impl/EngineImpl.java [hunk 5]
```diff
@@ -182,20 +192,25 @@ public JobId execute(OwsCode identifier,
...
         if (job != null) {
+            this.jobStatusLock.readLock().unlock();
```
Comment (autermann) [inline engine/src/main/java/org/n52/javaps/engine/impl/EngineImpl.java]: Just use `try {} finally {}`...

File: engine/src/main/java/org/n52/javaps/engine/impl/EngineImpl.java (hunk 2 above)
Comment (matthesrieke) [reply engine/src/main/java/org/n52/javaps/engine/impl/EngineImpl.java]: I was wondering, this is just for performance reasons, right?

File: engine/src/main/java/org/n52/javaps/engine/impl/EngineImpl.java (hunk 3 above)
Comment (matthesrieke) [reply engine/src/main/java/org/n52/javaps/engine/impl/EngineImpl.java]: strictly speaking, the `job.destroy()` is the problem here. But you are right. so a `try` in a `try`? =)

Comment (matthesrieke) [reply engine/src/main/java/org/n52/javaps/engine/impl/EngineImpl.java]: Looking at the code, switching the lines should be fine..

**Pull request merged**